### ISSUE TITLE
Secrets should not be stored in state

### ISSUE THREAD

**Issue by iloveicedgreentea** [opentofu/opentofu#801]
### OpenTofu Version

```shell
N/a
```

### Use Cases

There are tons of reasons to have TF generate secrets at deploy time - a database, a private key, etc. I am specifically not talking about providing a secret (e.g as a parameter, in the TF files). I am only referring to apply time generation of secrets, or use of things like a secrets store resource/data. Right now, TF will stubbornly store the secret in state even though is is completely unnecessary. This makes the state an unsecured secrets store which undermines all of the strict secrets manager controls.

### Attempted Solutions

N/a

### Proposal

TF should never store secrets in state, period. If a value absolutely must be compared (which seems unlikely), then a random token hash should be used (and specifically NOT a hash of the secret itself).

To know if a secret was generated or not, TF can simply store that boolean in the state. This way, it does not need to know what the secret is. An alternative solution is for the ignore_lifecycle block (which is currently a hacky way to change secrets post apply so the state secret is not valid) to not write the secret or secret has into state when it is set.

Essentially, we should have a clear method to not store secrets in state. Hashicorp has not managed to solve this in over nine years.

### References

- https://github.com/hashicorp/terraform/issues/516

**Comment by Yantrio** [comment]
I'm not saying that this does get solved by #297 but I do think there is a large overlap here.

In general these 2 issues together are creating a of an [XY-Problem](https://xyproblem.info/). And it may be a good idea for us to create a more general discussion around safe storage of sensitive information overall.

I'm happy to keep this open and discuss this individual idea as I feel it's important, but I do feel there is more of an over-arching story that is here too.

**Comment by iloveicedgreentea** [comment]
#297 is a great idea but it doesn't really solve the issue of TF simply not needing to store sensitive data in the first place.

**Comment by StamenchoBog** [comment]
I would also give my opinion on this topic. I completely agree with you, @iloveicedgreentea. I had this as my opinion: we really don't need to store secrets in the state files in TF. If this could be implemented, it would resolve a major security issue with the state file being exposed. The encryption solution, #297, isn't really a solution to this problem, just an additional security layer. But let's keep in mind that you cannot steal something that isn't there.

**Comment by zhenrong-wang** [comment]
In my practice, I developed an independent (external) text file encryption program to encrypt the state file after tf running, and decrypt the state file before next-time running.

I also use this method and the program to encrypt the .tf file with cloud credentials.

Although this doesn't eliminate the security risk (because when tf IS running, all the files have to be in plain text format), at least the sensitive information are not exposed when tf is not running.

**Comment by janosdebugs** [comment]
Hey folks, thank you very much for this issue and the related discussion. We discussed this issue in the core team and feel that the majority of the use cases where this would be used can be solved by #874 (state encryption), which we are currently working on. The rest of the use cases would be solved by moving sensitive values only to a key management system for storage, but that presents numerous issues, such as the lack of atomic changes and so on, which would likely result in large-scale architectural changes. As such, we will be closing this issue. If you have a use case that is not covered by state encryption and you cannot solve in any other way, please comment on that use case below.

**Comment by bgshacklett** [comment]
The biggest problem I see here is that the people who are responsible for managing infrastructure should rarely have access to the systems running on that infrastructure. To avoid crossing that barrier, one must not use Terraform to set, nor reference, any application level secrets, including things like DB passwords. So, a simple pattern like generating a secret using `random_password` and passing it to a secret store and a db instance must be avoided.

The encryption mechanism does not enable fine grained access, nor does it protect against internal threats. Anyone who has access to run Terraform against the system has access to the unencrypted state. This means one must add custom solutions on top of Terraform for any sensitive areas, and there is a lot of room for custom implementations of that nature to leak sensitive information.

**Comment by janosdebugs** [comment]
@bgshacklett given that OpenTofu/Terraform need to have access to the sensitive values to work, the person running `tofu` will always have access to these secrets, worst case with a debugger. The only solution I know of that mitigates this risk is not giving people access to production systems, but using a CI/CD system instead.

**Comment by bgshacklett** [comment]
I'm assuming that a CI system is in place. Arbitrary code can be run in CI. At the very least, it's quite easy to create a new output, using the `nonsensitive()` function to output the data, but there are other, less obvious, ways to dump the state as well. If the data is not stored in state, it's _only_ available in memory during the initial configuration. No, it's not perfect, but it increases the barrier to exfiltration. One would need to recreate the secret to get access to it, which is likely to set off a lot more alarm bells.

There will always be pathways to retrieval of secrets and other sensitive data, but we should not ignore steps that can be taken to make it harder and more likely to be caught, just because a more sophisticated attack is possible.

I still have not seen a comprehensive explanation of why a provider MUST be able to retrieve the original value from state, rather than retrieving and comparing a hash to decide whether a value needs to be updated. I understand that that is the current expectation, but creating an opt-in mechanism of this sort would allow providers to implement better security over time. Is that truly infeasible?

**Comment by janosdebugs** [comment]
@bgshacklett there are a number of pitfalls with the proposed approach. Assume the provider has no way to retrieve a secret value, but has an "update if different" function. This wouldn't work.

The main problem is that the provider SDK and protocol is in the hands of Hashicorp and a split would mean that hundreds if not thousands of provider authors need to build, test and ship providers for two systems. This would almost certainly put a huge dampener on the enthusiasm for OpenTofu for an edge case benefit.

**Comment by bgshacklett** [comment]
> The main problem is that the provider SDK and protocol is in the hands of Hashicorp and a split would mean that hundreds if not thousands of provider authors need to build, test and ship providers for two systems. This would almost certainly put a huge dampener on the enthusiasm for OpenTofu for an edge case benefit.

To be clear, I'm not about to advocate for an implementation that is going to create a rift in the ecosystem. As you pointed out, that's not going to be helpful to OpenTofu in the long term, and would void any benefit that was achieved. My hope is to find a way to handle this _without_ that split. Either by working with HC, or potentially finding a way to make this work without breaking compatibility. I'm aware there's some difficulty there, but I haven't seen much attempt at a possible solution, so ruling it out seems premature, even if it's not easy.

I really don't believe this is an edge case. I can easily see an argument being made that the current (and future encrypted) model runs afoul of PCI compliance requirements. Specifically separation of duties:
https://pcidssguide.com/what-is-the-separation-of-duties-principle-how-is-it-implemented/. 

In any case, I appreciate you taking the time to respond to my comments. I'll leave it alone unless I get an opportunity to work on a real technical solution.

**Comment by janosdebugs** [comment]
Hi @bgshacklett apologies for the late reply, I was out of office. The main issue is that providers currently have the expectation of being able to read the state even for data sources. Unless we are willing to break compatibility, we can't implement it.

As far as compliance is concerned, if you or any other community member has a pending compliance problem, please share specifically what compliance item is in need of a resolution in a new issue or via private channels so we can adequately address it.

**Comment by dmurvihill** [comment]
> Hey folks, thank you very much for this issue and the related discussion. We discussed this issue in the core team and feel that the majority of the use cases where this would be used can be solved by #874 (state encryption), which we are currently working on.
> 
> The rest of the use cases would be solved by moving sensitive values only to a key management system for storage, but that presents numerous problems, such as the lack of atomic changes and so on, which would likely result in large-scale architectural changes to OpenTofu. As such, we are not planning to implement this and will be closing this issue. If you have a use case that is not covered by state encryption and you cannot solve in any other way, please comment on that use case below.

I firmly disagree with this call. Encrypting an unnecessary copy of a secret does not solve the problem. You still have an unnecessary copy of a secret.

When I generate a database password and store it in AWS Secrets Manager, there is absolutely no reason for it to be stored anywhere other than there and in the virtual memory of the application that uses it. There is no reason for Terraform to use that value. Not for interpolation, not for difference checking, never.

This has been the overwhelming opinion on Terraform #516 for nearly ten years and a consistent source of frustration with Hashicorp. The fact that OpenTofu's position is even worse is very disappointing.

**Comment by zhenrong-wang** [comment]
> > Hey folks, thank you very much for this issue and the related discussion. We discussed this issue in the core team and feel that the majority of the use cases where this would be used can be solved by #874 (state encryption), which we are currently working on.
> > 
> > The rest of the use cases would be solved by moving sensitive values only to a key management system for storage, but that presents numerous problems, such as the lack of atomic changes and so on, which would likely result in large-scale architectural changes to OpenTofu. As such, we are not planning to implement this and will be closing this issue. If you have a use case that is not covered by state encryption and you cannot solve in any other way, please comment on that use case below.
> 
> I firmly disagree with this call. Encrypting an unnecessary copy of a secret does not solve the problem. You still have an unnecessary copy of a secret.
> 
> When I generate a database password and store it in AWS Secrets Manager, there is absolutely no reason for it to be stored anywhere other than there and in the virtual memory of the application that uses it. There is no reason for Terraform to use that value. Not for interpolation, not for difference checking, never.
> 
> This has been the overwhelming opinion on Terraform #516 for nearly ten years and a consistent source of frustration with Hashicorp. The fact that OpenTofu's position is even worse is very disappointing.

I support the openTofu's practice on state file encryption. It best protects secrets and credentials based on current technical restrictions. 

In my practice, I even developed an independent encryption module to protect the .tf files because cloud AK and SK or other secrets are in them.

**Comment by zhenrong-wang** [comment]
> @zhenrong-wang that's interesting. Why wouldn't you pass in the secrets using environment variables or inputs?

Well the point is, Envvars are in plain text files  and pretty easy for any application running locally to retrieve. That's much worse in terms of security compare to file encryption.

Setting AK and SK as global envvars are super bad and unacceptable.

Setting AK and SK as loadable, temporary envvars needs an envfile in plain text. E.g. /home/myname/.tfenv_credentials

This also equals to expose the credentials to all applications running locally. Of course, you can secure it by setting RW permissions, but the permission is file_based, not application based.

**Comment by dmurvihill** [comment]
> Hi @dmurvihill we discussed your comment with the core team and the main problem is that this would require a protocol change to implement. Currently, providers are allowed to and do rely on the data in the state. Removing it would break providers.
> 
> The protocol is currently in the hands of HashiCorp and we cannot modify it for OpenTofu only as it would split the ecosystem. As far as we can tell, HashiCorp are evaluating ephemeral support which would likely cause the requisite protocol change. If and when that happens, we'd be open to discussing ephemeral support in OpenTofu based on their protocol implementation. (The protocol implementation is still MPL-licensed.)

Thank you for this. Being blocked by the provider profile is a lot more understandable.

FWIW, HashiCorp has also been saying they are blocked by the provider protocol, even though they control it 🙄.
